codec-http2/src/main/java/io/netty/handler/codec/http2/Http2RemoteFlowController.java

```diff
@@ -36,24 +35,44 @@
      *
      * @param ctx the context from the handler.
      * @param stream the subject stream. Must not be the connection stream object.
-     * @param data payload buffer for the frame.
-     * @param padding the number of padding bytes to be added at the end of the frame.
-     * @param endOfStream Indicates whether this is the last frame to be sent to the remote endpoint
-     *            for this stream.
+     * @param payload payload to write subject to flow-control accounting and ordering rules.
      * @param promise the promise to be completed when the data has been successfully written or a
      *            failure occurs.
      * @return a future that is completed when the frame is sent to the remote endpoint.
      */
-    ChannelFuture sendFlowControlledFrame(ChannelHandlerContext ctx, Http2Stream stream,
-            ByteBuf data, int padding, boolean endStream, ChannelPromise promise);
+    ChannelFuture sendFlowControlledData(ChannelHandlerContext ctx, Http2Stream stream,
+          FlowControlledPayload payload , ChannelPromise promise);
 
     /**
-     * Gets the {@link ChannelFuture} for the most recent frame that was sent for the given stream
-     * via a call to {@link #sendFlowControlledFrame()}. This is useful for cases such as ensuring
-     * that {@code HEADERS} frames maintain send order with {@code DATA} frames.
-     *
-     * @param stream the subject stream. Must not be the connection stream object.
-     * @return the most recent sent frame, or {@code null} if no frame has been sent for the stream.
+     * Implementations of this interface are used to progressively write chunks of the underlying
+     * payload to the stream. A payload is considered to be fully written if {@link #write} has
+     * been called at least once and it's {@link #size} is now zero.
      */
-    ChannelFuture lastFlowControlledFrameSent(Http2Stream stream);
+    interface FlowControlledPayload {
```

Comment:
Agreed, used FlowControlled instead of FlowControlledData as it makes the per-type names more sensible.